### ISSUE TITLE
Blazegraph confuses strings with and without RTM (U+200F) 

### ISSUE THREAD

**Issue by smalyshev** [blazegraph/database#93]
If Blazegraph indexes a triple that has a string with RTM character (U+200F), e.g. "0000 0000 4698 056X\u200F", then it would consider it to be the same string as one without RTM, i.e. "0000 0000 4698 056X". For example, if a new triple is entered with non-RTM string, Blazegraph would still return the RTM string when the triple is returned. 

This seems to be happening because Blazegraph is using ICU collation keys as string keys, in https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/keys/ICUSortKeyGenerator.java, and by default ICU collator seems to be generating the same collation key for RTM and non-RTM strings.

**Comment by thompsonbry** [comment]
This is a matter of the collation strength. If it is set to IDENTICAL then
all differences are respected.

Bryan

**Comment by smalyshev** [comment]
There seems to be two issues here:

1. The default is set not to Identical (I think it may be Secondary? not sure) despite all Unicode being allowed in generic case
2. I don't see any config value that allows to change it. Maybe I am missing some options, but I could not find any path in code that leads to setting these. Is it documented somewhere maybe that I missed?

**Comment by thompsonbry** [comment]
https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/keys/DefaultKeyBuilderFactory.java

Especially

https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/keys/DefaultKeyBuilderFactory.java#L345

The Options interface declares the supported configuration properties.

See
https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/keys/KeyBuilder.java#L1639

Bryan

**Comment by smalyshev** [comment]
Yes, I found this code, but I do not see how it can ever be reached since DefaultKeyBuilderFactory us always called with either null argument or `new Properties()` in the upstream code. E.g. https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/DefaultTupleSerializer.java#L73

**Comment by thompsonbry** [comment]
https://github.com/blazegraph/database/blob/master/bigdata-core/bigdata/src/java/com/bigdata/btree/keys/DefaultKeyBuilderFactory.java#L168

You can use -DNAME=VALUE.  See below.

/**
* Return the property if found in <i>properties</i>. If <i>properties</i>
* is <code>null</code> or if the value is not found in <i>properties</i>,
* then return the property if found using
* {@link System#getProperty(String)}.
*
* @param properties
* The properties.
* @param key
* The name of the desired property.
* @param def
* The default (MAY be <code>null</code>).
*
* @return The value -or- <i>def</i> if no value was found.
*/
static private String getProperty(final Properties properties,
final String key, final String def) {
String val = null;
if (properties != null) {
val = properties.getProperty(key);//, def);
}
if (val == null) {
val = System.getProperty(key, def);
}
if(log.isDebugEnabled()) {
log.debug("name=" + key + ",val=" + val);
}
return val;
}

**Comment by smalyshev** [comment]
Ah, I did not notice it uses System properties too. I'll try that, thanks.

**Comment by thompsonbry** [comment]
No problem!

Bryan

**Comment by smalyshev** [comment]
Yes, it looks like setting `-Dcom.bigdata.btree.keys.KeyBuilder.collator.strength=Identical` fixes the issue, but keys become much longer. Also, seems to require reindexing the whole store from scratch. 

Will setting it to Identical have any performance impact?

**Comment by thompsonbry** [comment]
This will mainly hit dictionary operations.  The statement indices will be
unchanged, hence query performance will be essentially unchanged.  Yes, you
would need to reload since the Unicode sort keys stored in the forward
dictionary (Term2Id) would be different.

I think that it is a fair question as to what is the most appropriate level
of distinction.  This specific case of a RTM (Registered Trade Mark,
right?) could also be achieved by adding an assertion about an entity
having that value as a label.

Would you want to have all variations of case stored as distinct Literals?

Do you want distinctions in a sequence of code points for the same glyph to
be stored as different Literals?

Unicode is always tricky.

Bryan

**Comment by thompsonbry** [comment]
So, you can easily change what is in ID2TERM, which has the actual Unicode
string that is returned when a query result is materialized.  This would
let you strip out somethings that you might not want to have in there.  You
could then filter the data on insert (externally).

You could change the TERM2ID mapping also.  I am not sure what would happen
if you tried to have multiple keys mapped to the same ID in that index.  It
might be Ok. Or something might have a fit.

Just make sure that the server is not answering queries when making these
sorts of changes.  Otherwise you might well hit asserts otherwise or odd
behavior if the ID <=> TERM mapping was not fully consistent (including in
the TermCache maintained by the LexiconRelation, in the BigdataLiteral =>
IV => IVCache => BigdataLiteral relationships, etc.).

Bryan
